File: src/modules/algorithm/seed/algorithms/hamiltonian-cycle/typescript/backtracking-solution.ts
/**
 * Determines if a Hamiltonian Cycle exists in a given graph using backtracking.
 * @param {number[][]} graph - Adjacency matrix representing the graph.
 * @returns {number[] | null} - A Hamiltonian Cycle if it exists, otherwise null.
 */
export function findHamiltonianCycle(graph: number[][]): number[] | null {
  const numVertices = graph.length;
  const path: number[] = new Array(numVertices);
  const visited: boolean[] = new Array(numVertices).fill(false);

  // Start from vertex 0
  path[0] = 0;
  visited[0] = true;

  if (hamCycleUtil(graph, path, visited, 1)) {
    path[numVertices] = 0; // Complete the cycle
    return path;
  }

  return null;

  function hamCycleUtil(
    graph: number[][],
    path: number[],
    visited: boolean[],
    position: number,
  ): boolean {
    // Base case: all vertices are included in the path
    if (position === numVertices) {
      // Check if the last vertex is adjacent to the first vertex
      if (graph[path[position - 1]][path[0]] === 1) {
        return true;
      } else {
        return false;
      }
    }

    // Try different vertices as the next candidate in the Hamiltonian Cycle
    for (let vertex = 1; vertex < numVertices; vertex++) {
      if (graph[path[position - 1]][vertex] === 1 && !visited[vertex]) {
        path[position] = vertex;
        visited[vertex] = true;

        // Recur to construct the rest of the path
        if (hamCycleUtil(graph, path, visited, position + 1)) {
          return true;
        }

        // Backtrack: If adding the current vertex doesn't lead to a solution,
        // then remove it and try a different vertex
        visited[vertex] = false;
        path[position] = -1;
      }
    }

    // If no vertex can be added to the path, return false
    return false;
  }
}

File: src/modules/algorithm/seed/algorithms/hamiltonian-path/languages/go/backtracking-solution.go
package hamiltonianpath

// HamiltonianPath determines if a Hamiltonian path exists in the given graph using backtracking.
// graph represents the adjacency matrix of the graph.
// It returns true if a Hamiltonian path exists, false otherwise.
func HamiltonianPath(graph [][]int) bool {
	n := len(graph)
	if n == 0 {
		return false
	}

	path := make([]int, 0, n)
	visited := make([]bool, n)

	var findPath func(v int) bool
	findPath = func(v int) bool {
		path = append(path, v)
		visited[v] = true

		if len(path) == n {
			return true
		}

		for i := 0; i < n; i++ {
			if graph[v][i] == 1 && !visited[i] {
				if findPath(i) {
					return true
				}
			}
		}

		// Backtrack
		path = path[:len(path)-1]
		visited[v] = false
		return false
	}

	for startNode := 0; startNode < n; startNode++ {
		// Reset visited and path for the next start node
		path = path[:0]
		for i := range visited {
			visited[i] = false
		}
		if findPath(startNode) {
			return true
		}

	}

	return false
} 